libcudacxx/include/cuda/std/detail/libcxx/include/__algorithm/copy.h

```diff
@@ -0,0 +1,97 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCUDACXX___ALGORITHM_COPY_H
+#define _LIBCUDACXX___ALGORITHM_COPY_H
+
+#ifndef __cuda_std__
+#  include <__config>
+#endif // __cuda_std__
+
+#include "../__algorithm/unwrap_iter.h"
+#include "../__type_traits/enable_if.h"
+#include "../__type_traits/is_constant_evaluated.h"
+#include "../__type_traits/is_same.h"
+#include "../__type_traits/is_trivially_copy_assignable.h"
+#include "../__type_traits/remove_const.h"
+#include "../cstdlib"
+#include "../cstring"
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+template <class _InputIterator, class _OutputIterator>
+inline _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 _OutputIterator
+__copy(_InputIterator __first, _InputIterator __last, _OutputIterator __result)
+{
+  for (; __first != __last; ++__first, (void) ++__result)
+  {
+    *__result = *__first;
+  }
+  return __result;
+}
+
+#if !defined(_LIBCUDACXX_COMPILER_NVRTC)
+template <class _Tp, class _Up>
+inline _LIBCUDACXX_HIDDEN _LIBCUDACXX_HOST _LIBCUDACXX_CONSTEXPR_AFTER_CXX17 bool
+__host_memmove(_Up* __result, _Tp* __first, const size_t __n)
+{
+#if _LIBCUDACXX_STD_VER >= 20
+  if (_CUDA_VSTD::is_constant_evaluated())
+  {
+    return false;
+  }
+  else
+#endif // _LIBCUDACXX_STD_VER >= 20
+  {
+    _CUDA_VSTD::memmove(__result, __first, __n * sizeof(_Up));
+    return true;
+  }
+}
+#endif // !_LIBCUDACXX_COMPILER_NVRTC
+
+template <class _Tp,
+          class _Up,
+          __enable_if_t<_LIBCUDACXX_TRAIT(is_same, __remove_const_t<_Tp>, _Up), int> = 0,
+          __enable_if_t<_LIBCUDACXX_TRAIT(is_trivially_copy_assignable, _Up), int>   = 0>
+inline _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX17 _Up*
+__copy(_Tp* __first, _Tp* __last, _Up* __result)
+{
+  const size_t __n = static_cast<size_t>(__last - __first);
+  if (__n > 0)
+  {
+    // need to dispatch to a separate function to keep this sane
+#if !defined(_LIBCUDACXX_COMPILER_NVRTC)
+    NV_IF_TARGET(NV_IS_HOST, (if (__host_memmove(__result, __first, __n)) { return __result + __n; }));
```

Comment:
I am torn whether we should even try to optimize at all, because this could be a `host -> device` copy which would then break.

Maybe we just tear out all memcpy optimizations for now and add a comprehensive optimization at a later stage?